ambry-router/src/main/java/com/github/ambry/router/PutOperation.java

```diff
@@ -1109,6 +1115,54 @@ OperationTracker getOperationTrackerInUse() {
       return operationTracker;
     }
 
+    @Override
+    public boolean check() {
+      if (quotaChargeCallback == null || isCharged) {
+        return true;
+      }
+
+      return quotaChargeCallback.check();
+    }
+
+    @Override
+    public boolean charge() {
```

Comment:
As mentioned in the other PR, check & charge should be atomic operations.

Reply:
I want to take this up as a separate PR. The atomicity should actually come from the last layer of implementation which is QuotaSource. Will modify the call stack to reflect that.